BankLibrary/Bank.cs

```diff
@@ -5,22 +5,95 @@ namespace BankLibrary
 
         public void OpenAccount(OpenAccountParameters parameters)
         {
-            // TODO: check types compatibility
-            CreateAccount(parameters.AccountCreated, () => parameters.Type == AccountType.Deposit 
-                ? new DepositAccount(parameters.Amount) as T
-                : new OnDemandAccount(parameters.Amount) as T);
+            if ((parameters.Type == AccountType.Deposit && parameters.BankType == BankType.OnDemandAccount) || 
+                (parameters.Type == AccountType.OnDemand && parameters.BankType == BankType.DepositAccount))
+            {
+                throw new InvalidOperationException("An account with this type can not create");
+            }
+
+            CreateAccount(parameters.AccountCreated, () => parameters.Type == AccountType.Deposit
+                   ? new DepositAccount(parameters.Amount, parameters.Percentage) as T
+                   : new OnDemandAccount(parameters.Amount,  parameters.Percentage) as T);
+        }
+
+        public void ClosedAccount(CloseAccountParameters parameters)
+        {
+            AssertValidId(parameters.Id);
+            var account = _accounts[parameters.Id];
+            AddAndDeleteCloseEvent(parameters, account);
+            _accounts.Insert(parameters.Id, account);
+        }
+
+        public void PutAmount(PutAccountParameters parameters)
+        {
+            AssertValidId(parameters.Id);
+            var account = _accounts[parameters.Id];
+            AddAndDeletePutEvent(parameters, account);
+            _accounts.Insert(parameters.Id, account);
+        }
+
+        public void WithdrawMoney(WithdrawAccountParameters parameters)
+        {
+            AssertValidId(parameters.Id);
+            var account = _accounts[parameters.Id];
+            AddAndDeleteWithdrawEvent(parameters, account);
+            _accounts.Insert(parameters.Id, account);
+        }
+
+        public void IncrementDay()
+        {
+            for (int i = 0; i < _accounts.Count; i++)
+            {
+                _accounts[i].IncrementDays();
+                _accounts[i].CalculatePercentage();
+            }
         }
 
         private void CreateAccount(AccountCreated accountCreated, Func<T> creator)
         {
             var account = creator();
-            account.Open();
-            account.Created += accountCreated;
+            AddAndDeleteCreateEvent(accountCreated, account);
             _accounts.Add(account);
         }
+
+        private void AssertValidId(int id)
+        {
+            if (id < 0 || id >= _accounts.Count)
+            {
+                throw new InvalidOperationException("An account with this number does not exist");
+            }
+        }
+
+        private static void AddAndDeleteCreateEvent(AccountCreated accountCreated, T account)
+        {
+            account.Created += accountCreated;
+            account.Open();
+            account.Created -= accountCreated;
```

Comment:
a bit strange logic.. I expect it to be like
1. create account instance: `var account = new Account()`
2. Subscribe on all events:
```
account.Created += ...
account.Closed += ...
...
```
so create account params should include all NotifyXxx delegates
3. you can clean-up subscriptions when account closed:

```c#
private void CloseAccount(int id)
{
    var account = _accouns.Single(x => x.Id == id);
    account.Close();
    CleanupSubscriptions(account);
}

private void CleanupSubscriptions(Account account)
{
    account.Created = null;
    ...
}
```